src/stim/io/measure_record_reader.cc

```diff
@@ -312,63 +341,71 @@ size_t MeasureRecordReaderFormatR8::read_bytes(PointerRange<uint8_t> data) {
 }
 
 bool MeasureRecordReaderFormatR8::read_bit() {
-    if (position >= bits_per_record) {
-        throw std::out_of_range("Attempt to read past end-of-record");
+    if (!buffered_0s && !buffered_1s) {
+        buffer_data();
     }
-    if (generated_1s < run_length_1s) {
-        ++generated_1s;
-        ++position;
-        return true;
-    }
-    if (generated_0s < run_length_0s) {
-        ++generated_0s;
-        ++position;
+    if (buffered_0s) {
+        buffered_0s--;
+        position++;
         return false;
-    }
-    if (!update_run_length()) {
-        throw std::out_of_range("Attempt to read past end-of-file");
-    } else {
-        ++generated_1s;
-        ++position;
+    } else if (buffered_1s) {
+        buffered_1s--;
+        position++;
         return true;
+    } else {
+        throw std::invalid_argument("Read past end-of-record.");
     }
 }
 
 bool MeasureRecordReaderFormatR8::next_record() {
+    while (!is_end_of_record()) {
+        read_bit();
+    }
     position = 0;
-    return false;
+    have_seen_terminal_1 = false;
+    int i = fgetc(in);
+    if (i == EOF) {
+        return false;
+    }
+    ungetc(i, in);
+    return true;
 }
 
 bool MeasureRecordReaderFormatR8::is_end_of_record() {
-    if (position >= bits_per_record) {
-        return true;
-    }
-    if (generated_0s < run_length_0s) {
-        return false;
-    }
-    if (generated_1s < run_length_1s) {
-        return false;
-    }
-    return !update_run_length();
+    return position == bits_per_record && have_seen_terminal_1;
 }
 
-bool MeasureRecordReaderFormatR8::update_run_length() {
-    int r = getc(in);
-    if (r == EOF) {
-        return false;
+void MeasureRecordReaderFormatR8::buffer_data() {
+    assert(buffered_0s == 0);
+    assert(buffered_1s == 0);
+    if (is_end_of_record()) {
+        throw std::invalid_argument("Attempted to read past end-of-record.");
     }
-    run_length_0s = 0;
-    while (r == 0xFF) {
-        run_length_0s += 0xFF;
+
+    // Count zeroes until a one is found.
+    int r;
+    do {
         r = getc(in);
+        if (r == EOF) {
+            throw std::invalid_argument("r8 data hit end-of-file without seeing a 1 encoded at end of data.");
```

Comment:
This error message is inaccurate. It suggests that a 1 is required, but it is OK for an R8 file to encode a bitstring consisting of 0s only, e.g the following is a valid R8 file

```
0xFF 0x2D
```

which encodes 300 zeros and no ones. Similarly,

```
0x00
```

is a valid R8 file encoding the empty bitstring (and neither encodes nor uses any 1s!).

Suggestion:

```c++
throw std::invalid_argument("r8 file cannot be empty and must end with byte other than 0xFF");
```

Reply:
Made the error message clearer.

Reply:
Yeah, this is definitely better. I think the error may also be thrown on an empty file, but this is very minor (and I haven't followed the full codepath for this case).

Reply:
An empty file is invalid r8 data (assuming you expect a record to be present). I think this issue falls under having an explicit step to start a record being read.

Reply:
I agree empty file is invalid for r8 (hence its explicit appearance in suggestion above). This one is OK too though.